### PR TITLE
[9728] Discovery server v2.0

### PR DESCRIPTION
**Pull request by MiguelBarro**
Backport of new discovery server version 2 feature from master to 2.0.x (see #1501)

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2981/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/236/)
* Mac Build Status
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2991/)

File: src/cpp/rtps/writer/StatefulPersistentWriter.cpp [hunk 1]
```diff
@@ -51,12 +55,30 @@ void StatefulPersistentWriter::unsent_change_added_to_history(
...
+        const GUID_t& reader_guid,
+        const SequenceNumber_t& min_requested_sequence_number,
+        const SequenceNumber_t& max_requested_sequence_number,
+        const SequenceNumber_t& next_sequence_number)
+{
+    if (!log_error_printed_)
+    {
+        log_error_printed_ = true;
+        logError(RTPS_WRITER, "Inconsistent acknack received in Local Writer "
+                << writer_guid << ". Maybe the persistent database has been erased locally.");
```
Comment (IkerLuengo) [inline src/cpp/rtps/writer/StatefulPersistentWriter.cpp:76]: If the database has been erased locally, does the user need to do something? If so, I'd suggest the actions explicitly:
```suggestion
        logError(RTPS_WRITER, "Inconsistent acknack received in Local Writer "
                << writer_guid << ". If the persistent database has been erased locally, <do whatever needs to be done>.");
```
If there is no action to be taken by the user, and this is just a log to record the possibility, ignore this comment.

File: test/blackbox/BlackboxTestsDiscovery.cpp [hunk 2]
```diff
@@ -151,10 +151,11 @@ TEST(Discovery, StaticDiscovery)
...
 
     writer.history_kind(eprosima::fastrtps::KEEP_ALL_HISTORY_QOS).
-    durability_kind(eprosima::fastrtps::TRANSIENT_LOCAL_DURABILITY_QOS);
+            durability_kind(eprosima::fastrtps::TRANSIENT_LOCAL_DURABILITY_QOS);
     writer.static_discovery("PubSubWriter.xml").reliability(eprosima::fastrtps::RELIABLE_RELIABILITY_QOS).
-    unicastLocatorList(WriterUnicastLocators).multicastLocatorList(WriterMulticastLocators).
-    setPublisherIDs(1, 2).setManualTopicName(std::string("BlackBox_StaticDiscovery_") + TOPIC_RANDOM_NUMBER).init();
+            unicastLocatorList(WriterUnicastLocators).multicastLocatorList(WriterMulticastLocators).
+            setPublisherIDs(1,
+            2).setManualTopicName(std::string("BlackBox_StaticDiscovery_") + TOPIC_RANDOM_NUMBER).init();
```
Comment (IkerLuengo) [inline test/blackbox/BlackboxTestsDiscovery.cpp:158]:
```suggestion
            unicastLocatorList(WriterUnicastLocators).multicastLocatorList(WriterMulticastLocators).
            setPublisherIDs(1, 2).
            setManualTopicName(std::string("BlackBox_StaticDiscovery_") + TOPIC_RANDOM_NUMBER).init();
```

File: test/blackbox/BlackboxTestsDiscovery.cpp [hunk 3]
```diff
@@ -173,11 +174,12 @@ TEST(Discovery, StaticDiscovery)
...
-    history_kind(eprosima::fastrtps::KEEP_ALL_HISTORY_QOS).
-    durability_kind(eprosima::fastrtps::TRANSIENT_LOCAL_DURABILITY_QOS);
+            history_kind(eprosima::fastrtps::KEEP_ALL_HISTORY_QOS).
+            durability_kind(eprosima::fastrtps::TRANSIENT_LOCAL_DURABILITY_QOS);
     reader.static_discovery("PubSubReader.xml").
-    unicastLocatorList(ReaderUnicastLocators).multicastLocatorList(ReaderMulticastLocators).
-    setSubscriberIDs(3, 4).setManualTopicName(std::string("BlackBox_StaticDiscovery_") + TOPIC_RANDOM_NUMBER).init();
+            unicastLocatorList(ReaderUnicastLocators).multicastLocatorList(ReaderMulticastLocators).
+            setSubscriberIDs(3,
+            4).setManualTopicName(std::string("BlackBox_StaticDiscovery_") + TOPIC_RANDOM_NUMBER).init();
```
Comment (IkerLuengo) [inline test/blackbox/BlackboxTestsDiscovery.cpp:182]:
```suggestion
            unicastLocatorList(ReaderUnicastLocators).multicastLocatorList(ReaderMulticastLocators).
            setSubscriberIDs(3, 4).
            setManualTopicName(std::string("BlackBox_StaticDiscovery_") + TOPIC_RANDOM_NUMBER).init();
```

File: test/blackbox/BlackboxTestsDiscovery.cpp [hunk 4]
```diff
@@ -319,12 +321,13 @@ TEST(Discovery, ParticipantLivelinessAssertion)
     auto test_transport = std::make_shared<test_UDPv4TransportDescriptor>();
 
     reader.disable_builtin_transport().add_user_transport_to_pparams(test_transport).
-    lease_duration({ 0, 800000000 }, { 0, 500000000 }).reliability(eprosima::fastrtps::RELIABLE_RELIABILITY_QOS).init();
+            lease_duration({ 0, 800000000 },
+            { 0, 500000000 }).reliability(eprosima::fastrtps::RELIABLE_RELIABILITY_QOS).init();
```
Comment (IkerLuengo) [inline test/blackbox/BlackboxTestsDiscovery.cpp:325]:
```suggestion
            lease_duration({ 0, 800000000 }, { 0, 500000000 }).
            reliability(eprosima::fastrtps::RELIABLE_RELIABILITY_QOS).init();
```

**Review by IkerLuengo — CHANGES_REQUESTED** [review]

File: include/fastdds/rtps/history/WriterHistory.h [hunk 5]
```diff
@@ -33,28 +33,32 @@ class WriteParams;
  * Class WriterHistory, container of the different CacheChanges of a writer
  * @ingroup WRITER_MODULE
  */
-class WriterHistory : public History
+class WriterHistory : public rtps::History
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply include/fastdds/rtps/history/WriterHistory.h:36]: 👍

File: src/cpp/CMakeLists.txt [hunk 6]
```diff
@@ -194,6 +186,47 @@ set(${PROJECT_NAME}_source_files
...
+#   rtps/builtin/discovery/participant/PDPServer2.hpp
+    rtps/builtin/discovery/participant/PDPServer2.cpp
+#   rtps/builtin/discovery/participant/PDPServerListener2.hpp
+    rtps/builtin/discovery/participant/PDPServerListener2.cpp
+#   rtps/builtin/discovery/endpoint/EDPClient2.hpp
+    rtps/builtin/discovery/endpoint/EDPClient2.cpp
+#   rtps/builtin/discovery/endpoint/EDPServer2.hpp
+    rtps/builtin/discovery/endpoint/EDPServer2.cpp
+#   rtps/builtin/discovery/endpoint/EDPServerListeners2.hpp
+    rtps/builtin/discovery/endpoint/EDPServerListeners2.cpp
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply src/cpp/CMakeLists.txt:228]: 👍

File: src/cpp/fastrtps_deprecated/publisher/PublisherHistory.cpp [hunk 7]
```diff
@@ -415,10 +415,10 @@ bool PublisherHistory::is_key_registered(
...
-            (vit->second.cache_changes.empty() ||
-                (NOT_ALIVE_UNREGISTERED != vit->second.cache_changes.back()->kind &&
-                NOT_ALIVE_DISPOSED_UNREGISTERED != vit->second.cache_changes.back()->kind
-                )
-            )
+           (vit->second.cache_changes.empty() ||
+           (NOT_ALIVE_UNREGISTERED != vit->second.cache_changes.back()->kind &&
+           NOT_ALIVE_DISPOSED_UNREGISTERED != vit->second.cache_changes.back()->kind
+           )
+           )
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply src/cpp/fastrtps_deprecated/publisher/PublisherHistory.cpp:422]: 👍

File: src/cpp/rtps/builtin/BuiltinProtocols.cpp [hunk 8]
```diff
@@ -21,8 +21,9 @@
 #include <fastdds/rtps/common/Locator.h>
 
 #include <fastdds/rtps/builtin/discovery/participant/PDPSimple.h>
+// #include "./discovery/participant/PDPClient2.hpp"
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/BuiltinProtocols.cpp:24]: 👍

File: src/cpp/rtps/writer/StatefulPersistentWriter.cpp (hunk 1 above)
Comment (MiguelBarro) [reply src/cpp/rtps/writer/StatefulPersistentWriter.cpp:76]: 👍

File: test/blackbox/BlackboxTestsDiscovery.cpp (hunk 2 above)
Comment (MiguelBarro) [reply test/blackbox/BlackboxTestsDiscovery.cpp:158]: 👍

File: test/blackbox/BlackboxTestsDiscovery.cpp (hunk 3 above)
Comment (MiguelBarro) [reply test/blackbox/BlackboxTestsDiscovery.cpp:182]: 👍

File: test/blackbox/BlackboxTestsDiscovery.cpp (hunk 4 above)
Comment (MiguelBarro) [reply test/blackbox/BlackboxTestsDiscovery.cpp:325]: 👍

File: src/cpp/rtps/writer/StatefulPersistentWriter.cpp [hunk 9]
```diff
@@ -51,12 +55,30 @@ void StatefulPersistentWriter::unsent_change_added_to_history(
...
+        const GUID_t& reader_guid,
+        const SequenceNumber_t& min_requested_sequence_number,
+        const SequenceNumber_t& max_requested_sequence_number,
+        const SequenceNumber_t& next_sequence_number)
+{
+    if (!log_error_printed_)
+    {
+        log_error_printed_ = true;
+        logError(RTPS_WRITER, "Inconsistent acknack received in Local Writer "
+                << writer_guid << ". Beware that delete persistent database without reseting the matched readers prevents communication with them.");
```
Comment (IkerLuengo) [inline src/cpp/rtps/writer/StatefulPersistentWriter.cpp:76]:
```suggestion
                << writer_guid << ". Beware that deleting persistent database without resetting the matched readers prevents communication with them.");
```

**Review by IkerLuengo — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/writer/StatefulPersistentWriter.cpp (hunk 9 above)
Comment (MiguelBarro) [reply src/cpp/rtps/writer/StatefulPersistentWriter.cpp:76]: 👍

**Review by IkerLuengo — APPROVED** [review]
LGTM

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2997/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/244/)
* Mac Build Status
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3008/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3017/)
* Linux aarch64Build Status
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2704/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3028/)

File: include/fastdds/rtps/attributes/ServerAttributes.h [hunk 10]
```diff
@@ -26,12 +26,19 @@
...
-
 class ParticipantProxyData;
+} // fastrtps
+} // rtps
+
+namespace fastdds {
+namespace rtps {
+
+// To be eventually removed together with eprosima::fastrtps
+namespace aux = ::eprosima::fastrtps::rtps;
```
Comment (IkerLuengo) [inline include/fastdds/rtps/attributes/ServerAttributes.h:41]: Not needed anymore

**Review by IkerLuengo — CHANGES_REQUESTED** [review]

**Review by IkerLuengo — APPROVED** [review]
:+1:

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3022/)
* Linux aarch64Build Status
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2709/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3033/)

**Pull request merged**